File: Taller de Haskell/Problema_8/problema_8.py
# Solicita al usuario que ingrese su edad
# Convierte la entrada a un número entero
edad = int(input("Digite su edad: "))
# Define el precio base del boleto
boleto = 10000
# Evalúa el precio del boleto según la edad del usuario
if edad <= 2:
    # Niños de 2 años o menos no pagan boleto
    print("El precio del boleto es: $0")
elif edad <= 12:
    # Niños entre 3 y 12 años reciben un 45% de descuento
    descuento = boleto * 0.45
    precio_final = boleto - descuento
    print(f"El valor del boleto es de ${precio_final}")
elif edad <= 59:
    # Adultos entre 13 y 59 años pagan el precio completo
    descuento = 0
    precio_final = boleto - descuento
    print(f"El valor del boleto es de ${precio_final}")
else:
    # Adultos mayores de 60 años reciben un 35% de descuento
    descuento = boleto * 0.35
    precio_final = boleto - descuento
    print(f"El valor del boleto es de ${precio_final}")

File: Taller de Haskell/Problema_8/problema_8.hs
-- Importamos funciones necesarias para entrada/salida y validación
import System.IO (hFlush, stdout)
import Text.Read (readMaybe)

-- | Precio base del boleto (constante global)
boletoBase :: Float
boletoBase = 10000.0

-- | 'calcularPrecioBoleto' toma la edad y retorna el precio final del boleto.
calcularPrecioBoleto :: Int -> Float
calcularPrecioBoleto edad
    | edad <= 2   = 0.0                          -- Niños menores de 3 años no pagan
    | edad <= 12  = boletoBase * (1 - 0.45)      -- 45% de descuento
    | edad <= 59  = boletoBase                   -- Precio completo
    | otherwise   = boletoBase * (1 - 0.35)      -- 35% de descuento para mayores de 60

-- | 'leerInt' solicita un número entero desde la entrada estándar con validación
leerInt :: String -> IO Int
leerInt mensaje = do
    putStr mensaje
    hFlush stdout
    entrada <- getLine
    case readMaybe entrada of
        Just n  -> return n
        Nothing -> do
            putStrLn "Entrada inválida. Intente de nuevo."
            leerInt mensaje

-- | Función principal que solicita la edad e imprime el precio del boleto
main :: IO ()
main = do
    edad <- leerInt "Digite su edad: "
    let precio = calcularPrecioBoleto edad
    putStrLn $ "El valor del boleto es de: $" ++ show precio
